src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs

```diff
@@ -332,14 +333,36 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
             }
         }
 
-        public async ValueTask ReadExactlyAsync(Memory<byte> buffer, CancellationToken cancellationToken = default) =>
-            _ = await ReadAtLeastAsyncCore(buffer, buffer.Length, throwOnEndOfStream: true, cancellationToken).ConfigureAwait(false);
+        public ValueTask ReadExactlyAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            ValueTask<int> vt = ReadAtLeastAsyncCore(buffer, buffer.Length, throwOnEndOfStream: true, cancellationToken);
+            if (vt.IsCompletedSuccessfully)
+                return default;
+
+            // use the ValueTask<int>'s backing object to create a ValueTask without allocating here.
+            object? obj = vt._obj;
+            Debug.Assert(obj is Task || obj is IValueTaskSource);
+
+            return obj is Task task ?
+                new ValueTask(task) :
+                new ValueTask((IValueTaskSource)obj, vt._token);
```

Comment:
I'd suggest instead adding an `internal ValueTask DangerousAsValueTask()` method to `ValueTask<T>`.  Its implementation would be something like:
```C#
internal ValueTask DangerousAsValueTask()
{
    Debug.Assert(_obj is Task or IValueTaskSource);
    return new ValueTask(_object, _token, _continueOnCapturedContext);
}
```
and then this code in ReadExactlyAsync can just be:
```C#
return ReadAtLeastAsyncCore(...).DangerousAsValueTask();
```
...or something along those lines.

Reply:
I ended up creating a static `ValueTask.DangerousCreateFromTypedValueTask<TResult>` method because the ValueTask ctor that takes `object, short, bool` is `private`, and I didn't feel comfortable making it `internal` for all of CoreLib to see, and use by mistake. Thoughts?

Reply:
sounds good